File: challenge.py
############## IMPORT REQUIRED MODULES #########################

import requests
import re

###############  MAIN ##########################################

def main():
	
	print_header_yes = 1
	f = open("Challenge_data.vcf", "r");
	for line in f:
		#skip comments in VCF
		if line.startswith('#'):
			continue
		line = line.rstrip('\n');
		#split VCF fields by tab
		vcf_fields = line.split("\t")

		gene_details = construct_url_rest(vcf_fields[0], vcf_fields[1], vcf_fields[3], vcf_fields[4])
		type = get_most_deleterious_annotation(vcf_fields[7])
		depth = get_depth(vcf_fields[7])
		refcount = get_refcount(vcf_fields[7])	
		altcount = (depth - refcount)	
		refpercentage = refcount/depth
		altpercentage = 1-refpercentage
		exac_return = call_client(gene_details)
		print_output(type, depth, altcount, altpercentage, refpercentage, exac_return['allele_freq'], exac_return['variant_id'], exac_return['variant_in_gene'], print_header_yes);
		print_header_yes += 1


############### CONSTRUCT QUERY FOR EXAC DB #####################


def construct_url_rest(chromosome, position, ref, alt):
        return("-". join([chromosome, position, ref,alt]))


############### GET MOST DELETERIOUS ANNOTATION ##################

def get_most_deleterious_annotation(annotype):

        annotype = re.split("=",annotype)[-1];
        anno = []
	anno = re.split(",", annotype)
        if len(anno)==1:
		return anno[0]
        else:
                if filter(lambda x:'del' in x, anno) and filter(lambda x:'ins' in x, anno):
			return "indel"
                elif filter(lambda x:'del' in x, anno):
                        return "del"
                elif filter(lambda x:'ins' in x, anno):
                        return "ins"
                elif filter(lambda x:'complex' in x, anno):
                        return "complex"
                elif filter(lambda x:'mnp' in x, anno):
                        return "mnp";
                else:
			return "snp"


################## GET READ DEPTH ##############################
                
def get_depth(getdepth):

        getdepth = re.split(";", getdepth)[7]
        return int(re.split("=", getdepth)[1])


################ GET REFERENCE READ COUNTS #######################


def get_refcount(getrefcount):
	getrefcount = re.split(";", getrefcount)[28]
        return int(re.split("=", getrefcount)[1])


################# EXAC CONNECT ##################################

def call_client(gene_desc):
	result = requests.get("http://exac.hms.harvard.edu/rest/variant/%s" %  gene_desc)
        if result.status_code!=200:
                print 'No response for gene_desc';
        json_data = result.json()
	if 'allele_freq' in json_data['variant']:
		allele_freq = json_data['variant']['allele_freq']
	else:
		allele_freq = 0
	if 'variant_id' in json_data['variant']:
		variant_id = json_data['variant']['variant_id']
	else:
		variant_id = "No Exac Info"
	if 'genes' in json_data['variant']:
		variant_in_gene = json_data['variant']['genes']
	else:
		variant_in_gene = ["No Exac Info"]
	return {'allele_freq': allele_freq, 'variant_id': variant_id, 'variant_in_gene': variant_in_gene}



############### PRINT OUTPUT #####################################

def print_output(type, depth, altcount, altpercentage, refpercentage, allele_freq, variant_id, genes, print_header_yes):
	if print_header_yes==1:
		fmt = "{type:s}\t|\t{depth:s}\t|\t{altcount:s}\t|\t{altpercentage:s}\t|\t{refpercentage:s}\t|\t{allele_freq:s}\t|\t{variant_id:s}\t|\t{genes:s}"
		print fmt.format(type="----", depth="-----", altcount="-----", altpercentage="-----", refpercentage="-----", allele_freq="-------", variant_id="-----------", genes="---------------")
		print fmt.format(type="TYPE", depth="DEPTH", altcount="#VARS", altpercentage="%VARS", refpercentage="%REFS", allele_freq="EXAC_AF", variant_id="EXAC_VAR_ID", genes="GENE ANNOTATION")
		print fmt.format(type="----", depth="-----", altcount="-----", altpercentage="-----", refpercentage="-----", allele_freq="-------", variant_id="-----------", genes="---------------")
	else:
		fmt = "{type:s}\t|\t{depth:d}\t|\t{altcount:d}\t|\t{altpercentage:0.3f}\t|\t{refpercentage:0.3f}\t|\t{allele_freq:0.3f}\t|\t{variant_id:s}\t|\t{genes:s}"
		newgenes = "\t". join(map(str,genes))	
		print fmt.format(type=type, depth=depth, altcount=altcount, altpercentage=altpercentage, refpercentage=refpercentage, allele_freq=allele_freq, variant_id=variant_id, genes=newgenes)

if __name__ == '__main__':
    main()

File: challenge.pl
#!/usr/bin/perl -w

############ DECLARE MODULES ##################

use strict;
use Data::Dumper;
use lib "REST/lib/perl5/site_perl/REST";
use lib "JSON/lib/perl5/site_perl";
use Client;
use JSON;

########### DECLARE MAIN VARIABLES ############

my ($line, $gene_details, $type, $depth, $refcount, 
    $altcount, $refpercentage, $altpercentage, 
    $exac_return);

my @vcf_fields;

my $print_header_yes = 1;

########### MAIN ########################################

open(FH, "Challenge_data.vcf") or die "cannot open Challenge_data file for reading\n";

while($line=<FH>){
	
	#skip comments in VCF
	next if $line=~/^#/;
	chomp $line;

	#split fields by tab
	@vcf_fields = split /\t/, $line;
	
	$gene_details = construct_url_rest(($vcf_fields[0], $vcf_fields[1], $vcf_fields[3], $vcf_fields[4]));
	$type = get_most_deleterious_annotation($vcf_fields[7]);
	$depth = get_depth($vcf_fields[7]);
	$refcount = get_refcount($vcf_fields[7]);
	$altcount = $depth - $refcount;
	$refpercentage = $refcount/$depth;
	$altpercentage = 1-$refpercentage;
	$exac_return = call_client($gene_details);
	print_output($type, $depth, $altcount, $altpercentage, $refpercentage, $exac_return, $print_header_yes);
	$print_header_yes+=1;	

}
close FH;	

########## CONSTRUCT QUERY FOR EXAC DB #####################

sub construct_url_rest{
	
	my ($chromosome, $position, $ref, $alt) = @_;
        return($chromosome."-". $position."-". $ref. "-".$alt);
}

######### GET MOST DELETERIOUS ANNOTATION #####################

sub get_most_deleterious_annotation{
	
	my $annotype = (split /=/,$_[0])[-1];
	my @anno = split /,/, $annotype;
	if(scalar(@anno)==1){
		return $anno[0];
	}
	else{
		if (grep(/del/, @anno) && (grep /ins/, @anno)) {
			return "indel";
		}
		elsif(grep(/del/, @anno)){
			return "del";
		}
		elsif(grep(/ins/, @anno)){
			return "ins";
		}
		elsif(grep(/complex/, @anno)){
			return "complex";
		}
		elsif(grep(/mnp/, @anno)){
			return "mnp";
		}
		else{
			return "snp";
		}
	}

}

########### GET READ DEPTH ###################################

sub get_depth{

	my $getdepth = (split /;/, $_[0])[7];
        return ((split /=/, $getdepth)[1]);

}

########### GET REFERENCE READ COUNT ############################

sub get_refcount{

	my $getref = (split /;/, $_[0])[28];
	return ((split /=/, $getref)[1]);

}

########### EXAC CONNECT ########################################

sub call_client{
	my @client_return = ();
	my $gene = shift;
	my $client = REST::Client->new();
	$client->setHost("http://exac.hms.harvard.edu");
	$client->GET(
    	"/rest/variant/$gene"
	);
	
	my $response = from_json($client->responseContent());
	my $allele_freq = $response->{'variant'}->{'allele_freq'};
	
	# Set 0 for no allele frequency info

	if(!($allele_freq)){
		$allele_freq = 0;
	}

	my $variant_id = $response->{'variant'}->{'variant_id'};

	#Set NA for no variant id info from EXAC
	
	if(!($variant_id)){
                $variant_id = "No Exac Info";
        }
	
	my $variant_in_gene = $response->{'variant'}->{'genes'};

	if(!($variant_in_gene)){
		$variant_in_gene = ["No Exac Info"] ;
	}
	
	push(@client_return, $allele_freq);
	push(@client_return, $variant_id);
	push(@client_return, $variant_in_gene);

	return(\@client_return);
}

############## PRINT TOP 10 GENE INFO ###########################

sub print_gene_info{

	my $db_ref = shift;
        my $i; 

	for($i=0; $i<10; $i++){
                if($db_ref->[2]->[$i]){
                        print $db_ref->[2]->[$i];
                	print "\t";
		}
	}
	print "\n";
	return;
}

############# PRINT OUTPUT ######################################

sub print_output{
 	my ($type, $depth, $altcount,$altpercentage,$refpercentage, $exac_return, $print_header_yes) = @_;
	
	if($print_header_yes == 1){
		
		printf("%.8s\t|\t%.8s\t|\t%.8s\t|\t%.8s\t|\t%.8s\t|\t%.8s\t|\t%.50s\t|\t%.100s\n", "----", "-----", "-----", "----", "----", "-------", "-----------", "---------------");
		printf("%.8s\t|\t%.8s\t|\t%.8s\t|\t%.8s\t|\t%.8s\t|\t%.8s\t|\t%.50s\t|\t%.100s\n", "TYPE", "DEPTH", "\#VARS", "\%VAR", "\%REF", "EXAC_AF", "EXAC_VAR_ID", "GENE ANNOTATION");
		printf("%.8s\t|\t%.8s\t|\t%.8s\t|\t%.8s\t|\t%.8s\t|\t%.8s\t|\t%.50s\t|\t%.100s\n", "----", "-----", "-----", "----", "----", "-------", "-----------", "---------------");
		
	}	

	printf("%.8s\t|\t%.8s\t|\t%.8s\t|\t%.3f\t|\t%.3f\t|\t%.3f\t|\t%.50s\t|\t", $type, $depth, $altcount, $altpercentage, $refpercentage, $exac_return->[0], $exac_return->[1]); 
	print_gene_info($exac_return);
 
}
